multi_token_support_program/src/main.leo

```diff
@@ -0,0 +1,1202 @@
+        account: owner.account,
+        balance: authorized_balance.balance - remaining_burn_balance,
+        authorized_until: authorized_balance.authorized_until
+      };
+      authorized_balances.set(balance_key, new_authorized_balance);
+    }
+
+    // Update the token supply
+    let new_metadata: TokenMetadata = TokenMetadata {
+      token_id: owner.token_id,
+      name: token.name,
+      symbol: token.symbol,
+      decimals: token.decimals,
+      supply: new_supply,
+      max_supply: token.max_supply,
+      admin: token.admin,
+      external_authorization_required: token.external_authorization_required,
+      external_authorization_party: token.external_authorization_party
+    };
+    registered_tokens.set(owner.token_id, new_metadata);
+  }
+
+  async transition burn_private(input_record: Token, amount: u128) -> (Token, Future) {
+    assert(input_record.token_id != CREDITS_RESERVED_TOKEN_ID);
+    let output_record: Token = Token {
+        owner: input_record.owner,
+        amount: input_record.amount - amount,
+        token_id: input_record.token_id,
+        external_authorization_required: input_record.external_authorization_required,
+        authorized_until: input_record.authorized_until
+    };
+    return (output_record, finalize_burn_private(input_record.token_id, amount, self.caller));
+  }
+
+  async function finalize_burn_private(token_id: field, amount: u128, token_admin: address) {
+    // Check that the token exists, and that the caller is the token admin
+    // Check that the token supply - amount >= 0
+    let token: TokenMetadata = registered_tokens.get(token_id);
+    assert_eq(token.admin, token_admin);
+    let new_supply: u128 = token.supply - amount; // underflow will be caught by the VM
+
+    // Update the token supply
+    let new_metadata: TokenMetadata = TokenMetadata {
+      token_id: token_id,
+      name: token.name,
+      symbol: token.symbol,
+      decimals: token.decimals,
+      supply: new_supply,
+      max_supply: token.max_supply,
+      admin: token.admin,
+      external_authorization_required: token.external_authorization_required,
+      external_authorization_party: token.external_authorization_party
+    };
+    registered_tokens.set(token_id, new_metadata);
+  }
+
+  async transition prehook_public(
+    owner: TokenOwner,
+    amount: u128,
+    authorized_until: u32
+  ) -> Future {
+    return finalize_prehook_public(owner, amount, authorized_until, self.caller);
+  }
+
+  async function finalize_prehook_public(
+    owner: TokenOwner,
+    amount: u128,
+    authorized_until: u32,
+    caller: address
+  ) {
+    // Check that the caller has permission to authorize
+    let token: TokenMetadata = registered_tokens.get(owner.token_id);
+    let token_admin: address = token.admin;
+    let external_authorization_party: address = token.external_authorization_party;
+    assert(caller == token_admin || caller == external_authorization_party);
```

Comment:
Is there a usecase for allowing  `token_admin` to authorize in addition to `external_authorization_party`? Will it be enough to set external_authorization_party = admin during the registration?

Reply:
`assert(caller == external_authorization_party);`